File: Algorithm/2023.12/graywhales/표현가능한이진트리.java
/*
비트연산자 + 분할정복
비트연산할때 연산순서와 범위를 조심하자
*/
public class Solution {
    int n;

    public int[] solution(long[] numbers) {
        n = numbers.length;
        int[] answer = new int[n];
        for (int i = 0; i < n; i++) {
            if (numbers[i] == 1){ // 1은 되니까 추가
                answer[i] = 1;
            }else{
                answer[i] = check(numbers[i]);
            }
        }
        return answer;
    }
    private int check(long number) {
        for (int i = 5; i > 0; i--) {
            int root = (1<<i)-1;// mid값이 될 수 있는 bit 자릿수는 2^n -1 이다. 0,1,3,7,15,31
            
            if(i!=5&&number>=((long)1<<(root*2+1))) break; // number가 2^(mid*2+1) 보다 클 수는 없어 i=5일때는 long값 넘어가서 곤란
            
            if ((number & ((long)1<<root)) != 0 ){
                if(dc(number, i, root, true)) return 1;
            }
            
        }
        return 0;
    }

    private boolean dc(long number, int r, int now, boolean root) {//분할정복
        
        if (r == 0) //leaf는 신경안씀
            return true;
            
        int left = now + (1 << (r - 1)); //왼쪽 노드
        int right = now - (1 << (r - 1)); //오른쪽 노드
        
        boolean leftRoot = (number & (long)1 << left) != 0;
        boolean rightRoot = (number & (long)1 << right) != 0;

        if (!root && (leftRoot || rightRoot)) { //부모가 0인데 자식이 1이면 false
            return false;
        }

        return dc(number, r - 1, left, leftRoot) && dc(number, r - 1, right, rightRoot);

    }

}

File: Algorithm/2023.12/wolfy916/코딩테스트 기출문제/표현 가능한 이진트리.py
'''
표현 가능한 이진트리 - 프로그래머스 Lv. 3 (2023 KAKAO BLIND)
분류 : 트리
'''

def solution(numbers):
    
    # [A] 트리 순회 함수
    def check(l, m, r):
        nonlocal result
        visited[m] = True

        # [A-1] 좌측 서브트리
        left = (l + m) // 2
        if 0 <= left and not visited[left]:
            if bin_tree[m] == '0' and bin_tree[left] == '1':
                result = 0
                return
            check(l, left, m)

        # [A-2] 우측 서브트리
        right = (m + r) // 2
        if right < len(bin_tree) and not visited[right]:
            if bin_tree[m] == '0' and bin_tree[right] == '1':
                result = 0
                return
            check(m, right, r)
        

    answer = []
    for num in numbers:
        # [1] 포화이진트리의 노드수에 맞게 이진수 생성
        # ex) 8 = 0b1000 => 0001000
        bin_num = bin(num)[2:]
        k = 0
        cnt = 2 ** k
        while len(bin_num) > cnt:
            k += 1
            cnt += 2 ** k
        bin_tree = '0' * (cnt - len(bin_num)) + bin_num
        
        # [2] 루트노드부터 순회하며 만들수 있는 숫자인지 확인
        result = 1
        visited = [False] * len(bin_tree)
        check(0, len(bin_tree) // 2, len(bin_tree))
        answer.append(result)
        
    return answer